### PR TITLE
[AI] Add log dumper

### PR DESCRIPTION
**Pull request by Cjkjvfnby**
Rework log parser a bit. 

Many things could be improved, but now it works and I think it's ok to merge it as is. 

- I have reworked the mechanism of dumping the logs. Code is extracted to a separate module.
- Refactored parser, now it consistent so two stages, parse log to tokens,  and process tokens with the state machine. Looks a bit less complicated than when the previous version.
- Use [streamlit](https://streamlit.io/) +  [altair](https://altair-viz.github.io/) for visualization. It looks nice.

I think these changes are a good base for future investigations.

изображение

### CODE REVIEW

File: default/python/charting/README.md [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+Tool that parses the logs and build charts about AI
+
+## Pre-requirements
+```shell
+pip install requirements_charts.txt
+```
+
+## Run
+```shell
+streamlit run charting\streamlite_app.py
```
Comment (Grummel7) [inline default/python/charting/README.md:15]: My pip wanted `pip install -r requirements_charts.txt`
Also first command must be called from this directory, send from one further up and the path only works on windows. I suggest to remove `charting\`
Actually it must be run from the log directory... so perhaps:
```
cd <log directory>
streamlit run <path to charting>streamlite_app.py
```

**Review by Grummel7 — COMMENTED** [review]
Hm, didn't want to submit this comment as a review...

**Comment by Grummel7** [comment]
Strange: I got a half-working page (but without any graphs) once, but now I always get:
```
muzz@albatross:~/.local/share/freeorion$ streamlit run ~/freeorion-project/Grummel7/default/python/charting/streamlite_app.py 

  You can now view your Streamlit app in your browser.

  Local URL: http://localhost:8501
  Network URL: http://192.168.2.111:8501

2022-05-29 22:14:41.138 Traceback (most recent call last):
  File "/home/muzz/.local/lib/python3.9/site-packages/streamlit/scriptrunner/script_runner.py", line 475, in _run_script
    exec(code, module.__dict__)
  File "/home/muzz/freeorion-project/Grummel7/default/python/charting/streamlite_app.py", line 70, in <module>
    draw_plots(get_ais_data())
  File "/home/muzz/freeorion-project/Grummel7/default/python/charting/streamlite_app.py", line 64, in draw_plots
    configure_colors(ais_data)
  File "/home/muzz/freeorion-project/Grummel7/default/python/charting/streamlite_app.py", line 53, in configure_colors
    colors = sorted((item["player"], item["color"]) for item in ais_data)
  File "/home/muzz/freeorion-project/Grummel7/default/python/charting/streamlite_app.py", line 53, in <genexpr>
    colors = sorted((item["player"], item["color"]) for item in ais_data)
KeyError: 'player'
```

Perhaps we'll have to do a screen sharing session to analyse this.

**Comment by Cjkjvfnby** [comment]
> Perhaps we'll have to do a screen sharing session to analyse this.

have you played a game with this branch?  A couple of turns should be enough to populate logs.

**Comment by Grummel7** [comment]
> have you played a game with this branch? A couple of turns should be enough to populate logs.

I got it: You assume that it analyses a single user game. For testing the AI, most of the time I run multi-user games where I am only an observer, that's why it cannot identify a player.

**Comment by agrrr3** [comment]
> * Use [streamlit](https://streamlit.io/) +  [altair](https://altair-viz.github.io/) for visualization. It looks nice.

is that (streamlit) cloud based?

**Comment by Grummel7** [comment]
Got it working now, looks nice.
One idea: could you deduct the empire colours from the log file and use the same in the graphs?

**Comment by Cjkjvfnby** [comment]
> is that (streamlit) cloud based?

You could also run it locally.

**Comment by Grummel7** [comment]
Just got a division by zero error, after displaying a game in which one AI had been defeated:
`File "/home/muzz/freeorion-project/Grummel7/default/python/charting/extract_data.py", line 86, in process_output
    info["RP_Ratio"] = rp / pp`

File: default/python/charting/extract_data.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    (
+        turn,
+        rp,
+        pp,
+    ) = parse_output(value)
+    info = data["turns"][-1]
+    info["turn"] = turn
+    info["PP"] = pp
+    info["RP"] = rp
+    info["RP_Ratio"] = rp / pp
```
Comment (Grummel7) [inline default/python/charting/extract_data.py]:
```suggestion
    info["RP_Ratio"] = (rp / pp) if pp else 0
```

Comment (Cjkjvfnby) [reply default/python/charting/extract_data.py]: I'd prefer to drop it here. This kind of operation could be done on the presentation layer.

**Comment by Cjkjvfnby** [comment]
@Grummel7 , thank you for the feedback.  

I went ahead and merged it.  This I support scripts so we could be less strict with steps we are merging in.

Any requests are welcome.

**Pull request merged**